src/anki/notebuilder: allow filtering only lowest frequency

eyalmazuz opened the pull request:
Closes #202 

This feature extends the ``{frequencies}`` marker with the extended marker syntax and adds two new options: ``min-value`` which allows to export only the lowest value frequency out of all the options and ``value-only`` which strips the dictionary name and html list format to keep only raw numbers.

This allow memento to support card format that rely on raw frequency values to sort their cards.

Example of results:
``{frequencies}`` or ``{frequencies:min-value=false,value-only=false}``
![image](https://github.com/user-attachments/assets/a2f408ed-9115-4c56-b017-371ac5c527a3)


 ``{frequencies:min-value=true,value-only=false}``
![image](https://github.com/user-attachments/assets/522391bc-6804-42d2-82cf-19ddfa716856)


``{frequencies:min-value=false,value-only=true}`` (note: values are separated by ``\n``)
![image](https://github.com/user-attachments/assets/504f4969-f9ac-433d-9cdb-f9d41b897d9c)


``{frequencies:min-value=true,value-only=true}``
![image](https://github.com/user-attachments/assets/d71f661f-c1f2-4de7-adf2-2d8a3dfa201e)


Edit: I would add that this essentially serves the same end result of ``{frequency-harmonic-rank}`` and ``frequency-average-rank``, to sort anki cards base on frequency values, but by having different values. So this might not bring any novel/unique features to memento but allow for users to have more variety on some of the values/fields

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
  */
 [[nodiscard]]
-static QString buildFrequencies(const QList<Frequency> &frequencies)
+static QString buildFrequencies(const QList<Frequency> &frequencies, QHash<QString, QString> args)
```
Wider than 80 characters.
Don't pass by `args` by value, use a const reference.

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+    bool minValueOnly = false;
+
+    QString minFreqDictionary;
+    int minFreq = 9999999;
```
This magic number should be stored in a `constexpr` variable separate from this.

eyalmazuz replied:
added a constexpr

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+
+    if (args.contains(VALUE_ONLY_KEY))
+    {
+        valueOnly = args[VALUE_ONLY_KEY].toLower() == "true";
```
A better way to do this is using `QVariant::toBool`. Check line 356.

eyalmazuz replied:
I missed that, changed to ``QVariant::toBool``

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+
+    if (args.contains(MIN_VALUE_KEY))
+    {
+        minValueOnly = args[MIN_VALUE_KEY].toLower() == "true";
```
Use `QVariant`

eyalmazuz replied:
same here

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+            freqStr += "</li>";
+        }
+
+        if (minValueOnly && freq.freq.toInt() < minFreq)
```
Check for errors when doing `toInt`.

eyalmazuz replied:
I added a ``parseFrequencyValue`` method and added a default value there in case it failed parsing

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+            freqStr += freq.dictionary.endsWith(':') ? " " : ": ";
+        }
+
+        freqStr += freq.freq + (minValueOnly ? "" : "\n");
```
Using an if-statement rather than a ternary is more readable here since one option is to essentially do nothing.
HTML ignores newline characters. Make sure this accomplishes what you want.

eyalmazuz replied:
I changed ``\n`` to ``<br>`` so it'll properly break lines
and moved it to an if-statement

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+
+        if (minValueOnly && freq.freq.toInt() < minFreq)
+        {
+            minFreq = freq.freq.toInt();
```
Save this value in a variable so you only have to call this once. C++ has if-statement syntax that allows you to do
```cpp
if (int freqValue = freq.freq.toInt(); minValueOnly && freqValue < minFreq)
```
if you're worried about scope.

eyalmazuz replied:
since I added ``parseFrequencyValue`` there no need to parse the value here and no scope issues

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+        if (minValueOnly && freq.freq.toInt() < minFreq)
+        {
+            minFreq = freq.freq.toInt();
+            minFreqDictionary = freq.dictionary + (freq.dictionary.endsWith(':') ? " " : ": ");
```
80 characters

eyalmazuz replied:
fixed

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
 
-    if (freqStr == "<ul></ul>")
+
+    if (freqStr == "<ul></ul>" || freqStr == "")
```
Modifying this if-statement is unnecessary since the string is already empty.

eyalmazuz replied:
you are right, reverted it

ripose-jp commented on src/anki/notebuilder.cpp:
```diff
@@ -486,29 +486,88 @@ static QString getFuriganaPlain(const Term &term)
+    }
+
+
+    QString freqStr = "";
```
You don't need to assign this anything. The string is empty by default.

eyalmazuz replied:
I assumed it'll be more "readable" so I added it but I removed it to match the rest of the codebase

ripose-jp reviewed (changes requested):
I like this change overall, but one thing you're overlooking is that not all frequency dictionaries encode numbers or integers. For example, [VNFreq](https://mega.nz/file/CpcSBRQb#g8vAZyL8ClnTnTPqwdU5rM0eJeimJ0A9LL2Nci5QZ4w) encodes floating points from 0 to 100. Testing this out, this isn't support by Memento, but it should be, so I'll fix that later. An older edition of this dictionary used strings with 1 to 5 star characters. I don't really care how you handle these cases, but you should make a conscious choice.

Also make sure to document your markers on the UI help page.

eyalmazuz commented:
@ripose-jp 
I fixed the code based on your comments.
I added documentation to the markers in the UI help page.

> but one thing you're overlooking is that not all frequency dictionaries encode numbers or integers. For example, [VNFreq](https://mega.nz/file/CpcSBRQb#g8vAZyL8ClnTnTPqwdU5rM0eJeimJ0A9LL2Nci5QZ4w) encodes floating points from 0 to 100.

I converted the method to use double values so it could support future updates for memento and other dictionaries that don't use integers only.

regarding
> An older edition of this dictionary used strings with 1 to 5 star characters.

I couldn't find the older version of this dictionary
I did find:
https://docs.google.com/document/d/1IUWkvBxhoazBSTyRbdyRVk7hfKE51yorE86DCRNQVuw/edit?tab=t.0
which has version of frequency dictionaries with star format so I copied the star symbol from there

in [shoui dictionary collection](https://drive.google.com/drive/folders/1tTdLppnqMfVC5otPlX_cs4ixlIgjv_lH?usp=sharing) the VNFreq is V2 which uses int values from 1 to 35k similar to JPDB and others
same with [other collections](https://github.com/MarvNC/yomitan-dictionaries)

I would say that I created an helper function that tries to parse the string value of the frequency to double and if it fails it gives a default value of 1000.0.

but when the export happens, I use the original string value of the dictionary
so it will export "★★★★" for example instead of the literal value of 80.0 or 1000.0 (in case it fails to parse it)

ripose-jp commented:
I made an update to your branch. I found that returning 1000 for unusable values caused some problems. Particularly JPDB includes 1451 and 271907㋕ for 口を開けて. If `min-value` is set, 271907㋕ ends up getting used which doesn't seem right. I've changed it to only parse the numeric parts of the string and chuck the rest. If it can't get a numeric value, it just doesn't consider the value for the purposes of `min-value`. Let me know if this acceptable or if you have other ideas.

eyalmazuz commented:
> I made an update to your branch. I found that returning 1000 for unusable values caused some problems. Particularly JPDB includes 1451 and 271907㋕ for 口を開けて. If `min-value` is set, 271907㋕ ends up getting used which doesn't seem right. I've changed it to only parse the numeric parts of the string and chuck the rest. If it can't get a numeric value, it just doesn't consider the value for the purposes of `min-value`. Let me know if this acceptable or if you have other ideas.

I'm perplexed how you spotted that edge case(s). Honestly I don't think I have anything on top of my head that could be help improve it, I'm not really knowledgeable on the wide variety of dictionaries.
so I don't have any more comments.

ripose-jp commented:
>I'm perplexed how you spotted that edge case(s).

Just dumb luck skipping around Spirited Away. I can't think of any other problems with this PR, so I'm merging it.
